scheduler/src/cook/mesos/scheduler.clj

```diff
@@ -436,112 +444,237 @@
+                ;; the pending-jobs atom is repopulated
                 @(d/transact
                    conn
-                   (mapcat (fn [task-info]
-                             [[:job/allowed-to-start? (:job-id task-info)]
-                              {:db/id (d/tempid :db.part/user)
-                               :job/_instance (:job-id task-info)
-                               :instance/task-id (:task-id task-info)
-                               :instance/hostname (:hostname offer)
-                               :instance/start-time (now)
-                               ;; NB command executor uses the task-id
-                               ;; as the executor-id
-                               :instance/executor-id (get-in
-                                                       task-info
-                                                       [:executor :executor-id]
-                                                       (:task-id task-info))
-                               :instance/slave-id (:slave-id offer)
-                               :instance/progress 0
-                               :instance/status :instance.status/unknown
-                               :instance/preempted? false}])
-                           task-infos)))
-              (log/info "Matched offer" offer "to tasks" task-infos)
-              ;; This launch-tasks MUST happen after the above transaction in
-              ;; order to allow a transaction failure (due to failed preconditions)
-              ;; to block the launch
-              (meters/mark! scheduler-offer-matched (count (:ids offer)))
-              (meters/mark! matched-tasks (count task-infos))
-              (histograms/update! number-tasks-matched (count task-infos))
-              (meters/mark! matched-tasks-cpus (:cpus match-resource-requirements))
-              (meters/mark! matched-tasks-mem (:mem match-resource-requirements))
-              (mesos/launch-tasks
-                driver
-                (:ids offer)
-                (mapv #(dissoc % :job-id :num-ports) task-infos))))))
+                   (vec (apply concat upgrade-txns task-txns)))
+                (log/info "Launching" (count task-txns) "tasks")
+                (log/info "Upgrading" (count (:upgrade-backfill processed-matches)) "tasks from backfilled to proper")
+                (log/debug "Matched tasks" task-txns)
+                ;; This launch-tasks MUST happen after the above transaction in
+                ;; order to allow a transaction failure (due to failed preconditions)
+                ;; to block the launch
+                (meters/mark! scheduler-offer-matched
+                              (->> matches
+                                   (mapcat (comp :id :offer :leases))
+                                   (distinct)
+                                   (count)))
+                (meters/mark! matched-tasks (count task-txns))
```

Comment:
Well, we wouldn't necessarily have to go all the way back to datomic (could just rebuild the info based on the scheduler-contents), but the main point is, just adding these metrics into the function without reworking it would be rather ugly & complex.  Would be better served by reworking the function.